File: pst.ml
#!/usr/bin/ocaml
#load "unix.cma";;

type proc = { cmd  : string;
              pid  : int;
              ppid : int;
              uid  : int;
              args : string list; }

let escape cmd =
  let len = String.length cmd in
  let buf = Buffer.create (len + 8) in
  for i = 0 to len - 1 do
    let code = Char.code cmd.[i] in
    if cmd.[i] = '\\'
    then Buffer.add_string buf "\\\\"
    else if code > Char.code ' ' && code <= Char.code '~'
    then Buffer.add_char buf cmd.[i]
    else Printf.bprintf buf "\\%03o" code
  done;
  Buffer.contents buf

let arg_seq ch =
  let buf = Buffer.create 256 in
  let rec next () =
    match (try Some (input_char ch) with End_of_file -> None) with
      | None when Buffer.length buf = 0 -> None
      | None | Some '\000' ->
          let s = Buffer.contents buf in
          Buffer.clear buf;
          Some (s, ())
      | Some c ->
          Buffer.add_char buf c;
          next () in
  Seq.unfold next ()

let read_args ch =
  List.of_seq (arg_seq ch)

let with_in_channel name f =
  let ch = open_in name in
  try let res = f ch in close_in ch; res
  with e -> close_in ch; raise e

let build_proc pid =
  let dir = "/proc/" ^ string_of_int pid in
  let st = Unix.stat dir in
  let uid = st.Unix.st_uid in
  let args = with_in_channel (dir ^ "/cmdline") read_args in
  let proc_stat = with_in_channel (dir ^ "/stat") input_line in
  let index = String.index proc_stat '(' + 1 in
  let length = String.rindex proc_stat ')' - index in
  let cmd = String.sub proc_stat index length in
  let index = index + length + 4 in
  let length = String.length proc_stat - index in
  Scanf.sscanf (String.sub proc_stat index length) "%d"
    (fun ppid -> {cmd=cmd; args=args; pid=pid; ppid=ppid; uid=uid})

let get_pids dir =
  Array.fold_left
    (fun pids name ->
       try int_of_string name :: pids
       with Failure _ -> pids)
    [] (Sys.readdir dir)

let read_procs () =
  List.rev_map build_proc (get_pids "/proc")

let get_username =
  let cache = Hashtbl.create 32 in
  fun uid ->
    try Hashtbl.find cache uid
    with Not_found ->
      let name = (Unix.getpwuid uid).Unix.pw_name in
      Hashtbl.replace cache uid name;
      name

let show_tree start procs =
  let rec show_children depth more_at_depth parent last_uid =
    let children = List.filter (fun proc -> proc.ppid = parent) procs in
    let depth = depth + 1 in
    match List.rev (List.sort compare children) with
      | [] -> ()
      | last :: rest ->
          List.iter
            (show_proc depth false ((depth, true) :: more_at_depth) last_uid)
            (List.rev rest);
          show_proc depth true ((depth, false) :: more_at_depth) last_uid last
  and show_proc depth last more_at_depth last_uid proc =
    if depth <> 0 then
      begin
        for i = 1 to depth do
          let more = List.assoc i more_at_depth in
          print_string "  ";
          print_string
            (if i = depth
             then (if last then "`-" else "|-")
             else (if more then "| " else "  "))
        done
      end;
    Printf.printf "%s,%d%s%s\n"
      proc.cmd proc.pid
      (if proc.uid <> last_uid
       then "," ^ (get_username proc.uid)
       else "")
      (if List.length proc.args > 1
       then " " ^ String.concat " " (List.map escape (List.tl proc.args))
       else "");
    show_children depth more_at_depth proc.pid proc.uid in
  let root = List.find (fun proc -> proc.pid = start) procs in
  show_proc 0 true [] 0 root

let () = show_tree 1 (read_procs ())

File: pst.rs
use std::collections::{HashMap, HashSet};
use std::fs::{self, File};
use std::io::{self, BufRead, Read};
use std::os::unix::fs::MetadataExt;
use std::sync::OnceLock;

#[derive(Debug, PartialEq, PartialOrd)]
struct Process {
    cmd: String,
    pid: i32,
    ppid: i32,
    uid: u32,
    args: Vec<String>,
}

fn escape(cmd: &str) -> String {
    let mut buf = String::with_capacity(cmd.len() + 8);
    for c in cmd.chars() {
        match c {
            '\\' => buf.push_str("\\\\"),
            c if c > ' ' && c <= '~' => buf.push(c),
            c => buf.push_str(&format!("\\{:03o}", c as u32)),
        }
    }
    buf
}

fn split_args(content: &str) -> Vec<String> {
    content.strip_suffix('\0').unwrap_or(content)
        .split('\0')
        .map(String::from)
        .collect()
}

fn build_proc(pid: i32) -> io::Result<Process> {
    let proc_dir = format!("/proc/{}", pid);
    let metadata = fs::metadata(&proc_dir)?;
    let uid = metadata.uid();

    let cmdline_path = format!("{}/cmdline", proc_dir);
    let mut cmdline = String::new();
    File::open(cmdline_path)?.read_to_string(&mut cmdline)?;
    let args = split_args(&cmdline);

    let stat_path = format!("{}/stat", proc_dir);
    let mut stat = String::new();
    File::open(stat_path)?.read_to_string(&mut stat)?;

    let cmd_start = stat.find('(').unwrap_or(0) + 1;
    let cmd_end = stat.rfind(')').unwrap_or(stat.len());
    let cmd = stat[cmd_start..cmd_end].to_string();

    let stat_parts: Vec<&str> = stat[cmd_end + 4..].split_whitespace().collect();
    let ppid = stat_parts.get(0).and_then(|s| s.parse().ok()).unwrap_or(0);

    Ok(Process { cmd, args, pid, ppid, uid })
}

fn get_pids() -> io::Result<Vec<i32>> {
    let entries = fs::read_dir("/proc")?;
    let mut pids = Vec::new();
    for entry in entries {
        if let Ok(entry) = entry {
            if let Ok(pid) = entry.file_name().to_string_lossy().parse::<i32>() {
                pids.push(pid);
            }
        }
    }
    Ok(pids)
}

fn read_procs() -> io::Result<Vec<Process>> {
    get_pids()?.into_iter().map(build_proc).collect()
}

static USERNAME_MAP: OnceLock<HashMap<u32, String>> = OnceLock::new();

fn get_username(uid: u32) -> Option<String> {
    let username_map = USERNAME_MAP.get_or_init(|| {
        let mut map = HashMap::new();
        if let Ok(passwd_file) = File::open("/etc/passwd") {
            let reader = io::BufReader::new(passwd_file);
            for line in reader.lines().flatten() {
                let parts: Vec<&str> = line.split(':').collect();
                if parts.len() > 2 {
                    if let Ok(id) = parts[2].parse::<u32>() {
                        map.insert(id, parts[0].to_string());
                    }
                }
            }
        }
        map
    });
    username_map.get(&uid).cloned()
}

fn show_children(
    procs: &[Process],
    depth: i32,
    more_at_depth: HashSet<i32>,
    parent: i32,
    last_uid: u32
) -> io::Result<()> {
    let mut children: Vec<&Process> =
        procs.iter().filter(|p| p.ppid == parent).collect();
    if children.is_empty() {
        return Ok(());
    }
    let depth = depth + 1;
    children.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let last_child = children.pop().unwrap();
    for child in children {
        let mut more = more_at_depth.clone();
        more.insert(depth);
        show_proc(procs, depth, false, more, last_uid, &child)?;
    }
    let mut more = more_at_depth.clone();
    more.remove(&depth);
    show_proc(procs, depth, true, more, last_uid, &last_child)
}

fn show_proc(
    procs: &[Process],
    depth: i32,
    last: bool,
    more_at_depth: HashSet<i32>,
    last_uid: u32,
    proc: &Process
) -> io::Result<()> {
    let mut prefix = String::new();
    for i in 1..=depth {
        let more = more_at_depth.contains(&i);
        prefix.push_str("  ");
        if i == depth {
            prefix.push_str(if last { "└─" } else { "├─" });
        } else {
            prefix.push_str(if more { "│ " } else { "  " });
        }
    }
    let username = get_username(proc.uid).unwrap_or_else(|| proc.uid.to_string());
    let args = proc.args[1..].iter().map(|arg| escape(arg)).collect::<Vec<String>>().join(" ");
    println!(
        "{}{},{}{}{}",
        prefix,
        proc.cmd,
        proc.pid,
        if proc.uid == last_uid { "".to_string() } else { format!(",{}", username) },
        if args.is_empty() { "".to_string() } else { format!(" {}", args) },
    );
    show_children(procs, depth, more_at_depth, proc.pid, proc.uid)
}

fn show_tree(start: i32, procs: Vec<Process>) -> io::Result<()> {
    let root = procs.iter().find(|p| p.pid == start).unwrap();
    show_proc(&procs, 0, true, HashSet::new(), 0, root)
}

fn main() -> io::Result<()> {
    show_tree(1, read_procs()?)
}
